ui/app/adapters/kv/data.js

```diff
@@ -0,0 +1,75 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+import ApplicationAdapter from '../application';
+import { encodePath } from 'vault/utils/path-encoding-helpers';
+import { kvId } from 'vault/utils/kv-id';
+import { assert } from '@ember/debug';
+
+export default class KvDataAdapter extends ApplicationAdapter {
+  namespace = 'v1';
+
+  _urlForSecret(backend, path, version) {
+    const base = `${this.buildURL()}/${encodePath(backend)}/data/${encodePath(path)}`;
+    return version ? base + `?version=${version}` : base;
+  }
+
+  createRecord(store, type, snapshot) {
+    const { backend, path, version } = snapshot.record;
+    const url = this._urlForSecret(backend, path);
+    return this.ajax(url, 'POST', { data: this.serialize(snapshot) }).then((resp) => {
+      resp.id = kvId(backend, version, path);
+      return resp;
+    });
+  }
+
+  queryRecord(store, type, query) {
+    const { path, backend, version } = query;
+    return this.ajax(this._urlForSecret(backend, path, version), 'GET').then((resp) => {
+      resp.id = kvId(backend, version, path);
+      return resp;
+    });
+  }
+
+  findRecord(store, type, id) {
+    return this.ajax(`${this.buildURL()}/${id}`, 'GET').then((resp) => {
+      resp.id = id;
+      return resp;
+    });
+  }
+
+  /* Five types of delete operations */
+  deleteRecord(store, type, snapshot) {
+    const { backend, path } = snapshot.record;
+    const { deleteType, deleteVersions } = snapshot.adapterOptions;
+
+    if (!backend || !path) {
+      throw new Error('The request to delete or undelete is missing required attributes.');
+    }
+
+    switch (deleteType) {
+      case 'delete-latest-version':
+        return this.ajax(this._urlForSecret(backend, path), 'DELETE');
+      case 'delete-specific-version':
+        return this.ajax(this._urlForSecret(backend, path), 'POST', {
+          data: { versions: deleteVersions },
+        });
+      case 'destroy-specific-version':
+        return this.ajax(`${this.buildURL()}/${encodePath(backend)}/destroy/${encodePath(path)}`, 'PUT', {
```

Comment:
I thought about that. But this is the only time `destroy` `metadata` and `delete` are added to the url for this adapter. To me it read easier and didn't give the impression that these URLs were generated elsewhere, which is what I tend to imagine if the param is added to a function. Thoughts?

Reply:
Commenting for posterity. This may change. Just got off a call with design, and we might move the "destroy-everything" functionality to the metadata view tab and thus it would be moved to the metadata adapter. Destroy may also get duplicated as it's something a data and metadata user may be able to do. Undelete will stay here.